infrastructure/debugging.py

```diff
@@ -70,37 +77,67 @@ def setup_ptvsd(host: str = "localhost", port: int = 5678) -> bool:
                 }
             ]
         }
-    ]
+      ]
+    }
 
     :param host: host of the debug server
     :param port: port of the debug server
     :return: Whether debugger was initialized properly or not
     """
     succeeded = False
-    try:
-        import ptvsd
-
-        ptvsd.enable_attach((host, port))
-        succeeded = True
-    except Exception as e:
-        print("Unable to create ptvsd debugger: {}".format(e))
+    if _check_if_should_setup():
+        try:
+            import ptvsd
+
+            ptvsd.enable_attach((host, port))
+            succeeded = True
+        except Exception as e:
+            print("Unable to create ptvsd debugger: {}".format(e))
     return succeeded
 
 
 def setup_debugpy(host: str = "localhost", port: int = 5678) -> bool:
     """
-    Setup debugpy degugging service
+    Setup debugpy debugging service
+
+    Here is a sample VSCode configuration for connecting to the debug server in launch.json:
+
+    {
+    "version": "0.2.0",
+    "configurations": [
+        {
+            "name": "Python: Remote Attach",
+            "type": "python",
+            "request": "attach",
+            "connect": {
+                "host": "localhost",
+                "port": 5678
+            },
+            "pathMappings": [
+                {
+                    "localRoot": "${workspaceFolder}/<plugin_name>",
+                    "remoteRoot": "/Users/<user_name>/Library/Application Support/QGIS/QGIS3/profiles/edplanning/python/plugins/<plugin_name>"
+                }
+            ]
+        }
+      ]
+    }
 
     :param host: host of the debug server
     :param port: port of the debug server
     :return: Whether debugger was initialized properly or not
     """
     succeeded = False
-    try:
-        import debugpy
-
-        debugpy.listen((host, port))
-        succeeded = True
-    except Exception as e:
-        print("Unable to create debugpy debugger: {}".format(e))
+    if _check_if_should_setup() and not os.environ.get("QGIS_DEBUGPY_HAS_LOADED"):
+        try:
+            # extra guard for debugpy not to setup it twice
+            # (causes debugging session to hang)
+            os.environ["QGIS_DEBUGPY_HAS_LOADED"] = "1"
```

Comment:
This probably should be placed after debugpy.listen() so env variable won't be set up if exception is raised. Or maybe even better place would be else block of try-except.